source/game_sa/Entity/Ped/CPed.cpp

```diff
@@ -502,7 +503,151 @@ void CPed::UpdatePosition()
+            auto pStandingOnVehicle = reinterpret_cast<CVehicle*>(pStandingOnEntity);
+            if (pStandingOnVehicle->m_nVehicleClass == eVehicleClass::CLASS_BIG
+                && !pStandingOnVehicle->physicalFlags.bDestroyed) {
+
+                physicalFlags.bSubmergedInWater = false;
+                auto pSwimTask = m_pIntelligence->GetTaskSwim();
+                if (!pSwimTask)
+                    return;
+
+                pSwimTask->m_fSwimStopTime += CTimer::ms_fTimeStep;
+                return;
+            }
+        }
+    }
+
+    if (m_pPlayerData) {
+        const auto& vecPedPos = GetPosition();
+        float fCheckZ = vecPedPos.z - 3.0F;
+        CColPoint lineColPoint;
+        CEntity* pColEntity;
+        if (CWorld::ProcessVerticalLine(vecPedPos, fCheckZ, lineColPoint, pColEntity, false, true, false, false, false, false, nullptr)) {
+            if (pColEntity->m_nType == eEntityType::ENTITY_TYPE_VEHICLE) {
+                auto pColVehicle = reinterpret_cast<CVehicle*>(pColEntity);
+                if (pColVehicle->m_nVehicleClass == eVehicleClass::CLASS_BIG
+                    && !pColVehicle->physicalFlags.bDestroyed
+                    && pColVehicle->GetMatrix()->GetUp().z > 0.0F) {
+
+                    physicalFlags.bSubmergedInWater = false;
+                    return;
+                }
+            }
+        }
+    }
+
+    // Those 4 are called for some reason
+    CTimeCycle::GetAmbientRed();
+    CTimeCycle::GetAmbientGreen();
+    CTimeCycle::GetAmbientBlue();
+    rand();
+	// Add splash particle if it's the first frame we're touching water, and
+	// the movement of ped is downward, preventing particles from being created
+	// if ped is standing still and water wave touches him
+    if (!physicalFlags.bTouchingWater && m_vecMoveSpeed.z < -0.01F) {
+        const auto& vecPedPos = GetPosition();
+        auto vecMoveDir = m_vecMoveSpeed * CTimer::ms_fTimeStep * 4.0F;
+        auto vecSplashPos = vecPedPos + vecMoveDir;
+        float fWaterZ;
+        if (CWaterLevel::GetWaterLevel(vecSplashPos.x, vecSplashPos.y, vecSplashPos.z, &fWaterZ, true, nullptr)) {
+            vecSplashPos.z = fWaterZ;
+            g_fx.TriggerWaterSplash(vecSplashPos);
+            AudioEngine.ReportWaterSplash(this, -100.0F, 1U);
+        }
+    }
+
+    physicalFlags.bTouchingWater = true;
+    physicalFlags.bSubmergedInWater = true;
+    ApplyMoveForce(vecBuoyancy);
+
+    const auto& vecPedPos = GetPosition();
+    if (CTimer::ms_fTimeStep / 125.0F < vecBuoyancy.z / m_fMass
+        || vecPedPos.z + 0.6F < mod_Buoyancy.m_fWaterLevel) {
+
+        bIsStanding = false;
+        bIsDrowning = true;
+
+        bool bPlayerSwimmingOrClimbing = false;
+        int iUnused;
+        if (!IsPlayer()) {
+           
+            CEventInWater cEvent(0.75F);
+            iUnused = 1;
+            m_pIntelligence->m_eventGroup.Add(&cEvent, false);
+            iUnused = -1;
```

Comment:
```suggestion
            CEventInWater cEvent(0.75F);
            GetEventGroup().Add(&cEvent, false);
```
This looks like some SEH code, you can remove it.